File: credit/src/main/java/ru/patterns/credit/infrastructure/metric/Counter.java
package ru.patterns.credit.infrastructure.metric;

import ru.patterns.credit.shared.kafka.MetricDto;

import java.time.Instant;

public interface Counter {

    static Builder builder(String metricName) {
        return new Builder(metricName);
    }

    final class Builder {
        private final String metric;

        Builder(String metric) {
            this.metric = metric;
        }


        public void register(MetricRegistry registry) {
            registry.register(new MetricDto(
                    metric,
                    Instant.now().toEpochMilli()
            ));
        }
    }
}

File: core/src/main/kotlin/ru/patterns/core/utils/metric/Counter.kt
package ru.patterns.core.utils.metric

import ru.patterns.core.utils.metric.serialization.MetricDto
import java.time.Instant
import java.util.UUID

interface Counter {
    companion object {
        fun builder(metricName: String): Builder = Builder(metricName)

        class Builder(
            private val name: String
        ) {
            private var requestId: String? = null

            fun requestId(requestId: String): Builder {
                this.requestId = requestId
                return this
            }

            fun register(metricRegistry: MetricRegistry) {
                metricRegistry.register(
                    MetricDto(
                        requestId = requestId ?: UUID.randomUUID().toString(),
                        metricName = name,
                        timestamp = Instant.now().toEpochMilli()
                    )
                )
            }
        }
    }
}